bin/run_docker.py

```diff
@@ -42,6 +42,51 @@ def get_submission_image(syn: synapseclient.Synapse, submission_id: str) -> str:
     return image_id
 
 
+def make_invalid_output(file_name, log_file_path, file_content):
+
+    dummy_file_name = "INVALID_"+file_name
+    dummy_file = os.path.join(log_file_path, dummy_file_name)
+    with open(dummy_file, "w") as d:
+        d.write(file_content)
+
+    return dummy_file
+
+
+def handle_outputs(output_path: str, output_file_name: str, log_text: str):
+    """
+    Handles any output files generated by the submitted Docker container:
+    Checks if any output files were generated by the container
+    run. If any were found, it will return the path to the first one found.
+    If none were found, it will return None.
+
+    Arguments:
+        output_path: The path to the output directory.
+        output_file_name: The name of the output file.
+        log_text: The log text to be updated, if an incorrect number of
+                  output files were found.
+
+    Returns:
+        output_file: The path to the output file.
+        log_text: The log text that was updated, if an incorrect number of
+                  output files were found.
+
+    """
+    # Glob any expected output files stored in the output_dir as a result of the container run
+    file_glob = glob(os.path.join(output_path, output_file_name+".*"))
+
+    # Update the log file if there is a case where too many output files were generated, or none were generated
+    if len(file_glob) != 1:
+        no_output_msg = f"Expected 1 Docker container output file in the output directory. Got {len(file_glob)}. If multiple output files were generated, the first will be used for validation and scoring."
```

Comment:
@BWMac for the edge-case where a participant's Docker container generates more than 1 output file, I was thinking of adding logic that retrieves and renames only the first file that gets `glob`ed in `run_docker.py`, and having the renaming function include the letters "pp" for "post-process", which we can then use in `run_docker.nf` for generating the predictions file output channels.

So for example:
1. Submission 123 generates 2 predictions files
2. In `run_docker.py` the first predictions file gets pulled and renamed `123_pp_predictions.csv`
3. In `run_docker.nf` the regex used in the predictions file output channel is `*pp*predictions.{csv, zip}` which distinguishes this output file from the rest, as it will be the only one validated and scored.

Reply:
I'm not sure we need to do that much here. If the guideline for a challenge states that models should only produce one file with the predictions name, I think it's fine to just have that submission fail validation if more than one file is produced and tell the submitter why. Accepting the first file arbitrarily and then continuing through to scoring seems like it could be more confusing overall.

Reply:
And as far as `UPDATE_FOLDERS` is concerned, you could just include some logic to not upload the predictions file if it matches a dummy file name/does not match a "correct" file naming convention

Reply:
That was my first thought but from an engineering standpoint it felt kind of sloppy, since `update_folders.py` would then depend on a very specific piece of logic in `run_docker.py` that may or may not exist in the long-term. Moreover, I like the idea of users having an "INVALID" predictions file that keeps a record of what submissions returned an incorrect number of outputs.

Reply:
I agree it isn't the cleanest thing in the world, but `update_folders.py` would not strictly depend on `run_docker.py`. I would argue that checking the name of a file before uploading would not remove any functionality and `update_folders.py` could still be used in other conforming contexts.

That said, I think it's fine not to do it that way. How are you thinking of preventing the dummy files from being uploaded (or are you at all)?

Reply:
I was originally planning to let the file be uploaded, since it has more useful information about why the submission is invalid, but I can also just add this information in the e-mail that gets sent out. I think this is one of those things that would work fine either way, and I'm not attached to one idea or the other, so if you feel strongly about not uploading the invalid file I can add that into `update_folders.py`.

<img width="702" alt="image" src="https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/assets/32107699/21136c77-539a-411e-bf5b-78e28c3a42bb">

Reply:
Ah, I think that is another point of confusion. Not sure if there will be differences based on challenge, but for the dynamic challenge at least, the email sent at the end includes a link to a wiki page where the submitter can see a "slice" of the submission view with only their results, including that information.

![Screenshot 2024-04-03 at 1 40 29 PM](https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/assets/52762200/80aaf7ab-8a3e-495a-a5fc-96185d51dde5)

Reply:
I see - It might be a good idea then to loop in Verena and Gaia in the discussion. I'll share the link to this thread with them.

Reply:
Based on the discussion here and with Verena & Gaia, the new handling of the output files includes:

* Returning a validation error to users when 0 OR >1 output files are produced. We should not quietly validate/score the first output file retrieved, if there is >1.
* The INVALID predictions file that gets generated for submissions that produce an incorrect number of outputs does _not_ get uploaded into Synapse.

I'm going to do a tower run to test that this all works as expected, and will resolve this thread when finished.

Reply:
[Passed!](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/4ubyDWEhpZKjQp)

No INVALID predictions file uploaded to Synapse

<img width="341" alt="image" src="https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/assets/32107699/3d00405f-4f53-442d-aee7-c83dc2915570">

But Docker execution log contains the error message

<img width="767" alt="image" src="https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/assets/32107699/c79d9e7a-d206-4d15-ac15-0c25160af858">

And so does e-mail

<img width="825" alt="image" src="https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/assets/32107699/50240283-ffef-49fd-9941-7ccfb10ce1cf">

I'll be resolving this thread now. Thanks for the helpful discussion @BWMac !